src/Native/Unix/System.Native/pal_console.c

```diff
@@ -71,92 +73,200 @@ void SystemNative_SetKeypadXmit(const char* terminfoString)
+static struct termios g_initTermios;          // the initial attributes captured
+
+static bool g_hasCurrentTermios = false;      // tracks whether g_currentTermios is valid
+static struct termios g_currentTermios;       // the latest attributes set
+
+// The terminal can be used by the .NET application via the Console class.
+// It may also be used by child processes that is started via the Process class.
+// The terminal needs to be configured different depending on the user.
+// ConfigureTerminalForXXX are called to change the configuration.
+// When it is ambiguous whether we should configure for Console/a child Process,
+// we prefer configuring for the Console.
+static bool g_reading = false;                // tracks whether the application is performing a Console.Read operation
+static bool g_childUsesTerminal = false;      // tracks whether a child process is using the terminal
+static bool g_terminalUninitialized = false;  // tracks whether the application is terminating
+
+static bool g_noTty = false;                  // cache we are not a tty
+
+static volatile bool g_receivedSigTtou = false;
+
+static void ttou_handler(int signo)
 {
-    // pal_signal.cpp calls this on SIGQUIT/SIGINT.
-    // This can happen when SystemNative_InitializeConsole was not called.
-
-    // Put the attributes back to what they were when the console was initially initialized.
-    // We only do so, however, if we have explicitly modified the termios; doing so always
-    // can result in problems if the app is in the background, as then attempting to call
-    // tcsetattr on STDIN_FILENO will suspend the app and prevent its shutdown. We also don't
-    // want to, for example, just compare g_currTermios with g_initTermios, as we'd then be
-    // factoring in changes made by other apps or by user code.
-    if (g_haveInitTermios &&                     // we successfully initialized the console
-        (g_readInProgress || !g_signalForBreak)) // we modified attributes
+    (void)signo;
+    g_receivedSigTtou = true;
+}
+
+static void InstallTTOUHandler(void (*handler)(int))
+{
+    struct sigaction action;
+    memset(&action, 0, sizeof(action));
+    action.sa_handler = handler;
+    int rvSigaction = sigaction(SIGTTOU, &action, NULL);
+    assert(rvSigaction == 0);
+    (void)rvSigaction;
+}
+
+static bool TcSetAttr(struct termios* termios, bool blockIfBackground)
+{
+    if (g_terminalUninitialized)
     {
-        tcsetattr(STDIN_FILENO, TCSANOW, &g_initTermios);
-        // ignore any failure
+        // The application is exiting, we mustn't change terminal settings.
+        return true;
     }
+
+    if (!blockIfBackground)
+    {
+        // When the process is running in background, changing terminal settings
+        // will stop it (default SIGTTOU action).
+        // We change SIGTTOU to to get EINTR instead of blocking.
+        InstallTTOUHandler(ttou_handler);
+
+        g_receivedSigTtou = false;
+    }
+
```

Comment:
Is it possible a signal handler could run here?  IIRC, some of our signal handlers do writes; if they get EINTR, will they end up in an infinite retry loop?

Reply:
Yes. When the sigttou signal handler is installed, and the terminal is configured with TOSTOP (not default), then a write to the terminal will spin on EINTR until the handler is removed.
The retry should stop when the handler is removed, and then the process should be stopped by the default sigttou handler on the next write attempt.

Reply:
What happens if this thread ends up processing the signal though?

Reply:
I'm not sure what you are thinking of.
We can add `SA_RESETHAND` when installing the sigttou handler. So if a second signal is generated (by a `write`/`tcsetattr` to the terminal), the process will be stopped, and there will be no spinning.

Reply:
> I'm not sure what you are thinking of.

Signals can be delivered on any thread, right? So if here we set it so that a write will result in EINTR, and then a signal arrives and a signal handler that's run here on this thread ends up calling write, it would spin forever in an EINTR retry loop because it's the code on this thread after the signal handler completes that would understand the change. Is that not possible?

Reply:
> Signals can be delivered on any thread, right? So if here we set it so that a write will result in EINTR, and then a signal arrives and a signal handler that's run here on this thread ends up calling write, it would spin forever in an EINTR retry loop because it's the code on this thread after the signal handler completes that would undo the change. Is that not possible?

Ah, yes, that is possible! The `SA_RESETHAND` will solve that too.

Reply:
I've added `SA_RESETHAND`.

Reply:
> The SA_RESETHAND will solve that too.

Can you help me understand how it addresses it?

Reply:
> Can you help me understand how it addresses it?

If something performs a `write` after we've installed the handler, `SA_RESETHAND` causes the default to be restored.
So when that `write` returns `EINTR`, and we re-try the write, the default handler causes the app to be stopped (as if we had never installed the custom handler).

Reply:
I see.  Ok.